pkg/manager/suspender/suspender.go

```diff
@@ -0,0 +1,244 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package suspender
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	errutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/klog/v2"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+)
+
+var (
+	suspendOrderForTC = []v1alpha1.MemberType{
+		v1alpha1.TiDBMemberType,
+		v1alpha1.TiFlashMemberType,
+		v1alpha1.TiCDCMemberType,
+		v1alpha1.TiKVMemberType,
+		v1alpha1.PumpMemberType,
+		v1alpha1.PDMemberType,
+	}
+	suspendOrderForDM = []v1alpha1.MemberType{
+		v1alpha1.DMWorkerMemberType,
+		v1alpha1.DMMasterMemberType,
+	}
+
+	_ Suspender = &suspender{}
+	_ Suspender = &FakeSuspender{}
+)
+
+type Suspender interface {
+	// SuspendComponent suspends the component if needed.
+	//
+	// Returns true if the component is needed to be suspended, and the reconciliation should be skipped.
+	SuspendComponent(v1alpha1.Cluster, v1alpha1.MemberType) (bool, error)
+}
+
+func NewSuspender(deps *controller.Dependencies) Suspender {
+	return &suspender{
+		deps: deps,
+	}
+}
+
+type suspendComponentCtx struct {
+	cluster   v1alpha1.Cluster
+	component v1alpha1.MemberType
+
+	spec   v1alpha1.ComponentAccessor
+	status v1alpha1.ComponentStatus
+}
+
+func (c *suspendComponentCtx) ComponentID() string {
+	return fmt.Sprintf("%s/%s:%s", c.cluster.GetNamespace(), c.cluster.GetName(), c.component)
+}
+
+type suspender struct {
+	deps *controller.Dependencies
+}
+
+func (s *suspender) SuspendComponent(cluster v1alpha1.Cluster, comp v1alpha1.MemberType) (bool, error) {
+	ctx := &suspendComponentCtx{
+		cluster:   cluster,
+		component: comp,
+		spec:      cluster.ComponentSpec(comp),
+		status:    cluster.ComponentStatus(comp),
+	}
+	if ctx.spec == nil || ctx.status == nil {
+		return false, fmt.Errorf("spec or status for component %s is not found", ctx.ComponentID())
+	}
+
+	suspending := cluster.ComponentIsSuspending(comp)
+
+	if !needsSuspendComponent(ctx.cluster, ctx.component) {
+		if suspending {
+			err := s.end(ctx)
+			return true, err
```

Comment:
Why return true here?

Reply:
because phase is still `Suspend` before tc update. It won't arrive here after the phase is updated.

Reply:
OK.